src/libraries/Native/Unix/System.Native/pal_process.c

```diff
@@ -871,3 +867,60 @@ int32_t SystemNative_SchedGetAffinity(int32_t pid, intptr_t* mask)
     return result;
 }
 #endif
+
+// Returns the full path to the executable for the current process resolving symbolic links.
+// The caller is responsible for releasing the buffer. Returns null on error.
+char* SystemNative_GetProcessPath()
+{
+    // Get path to the executable for the current process using
+    // platform specific means.
+#if defined(__APPLE__)
+    // On Mac, we ask the OS for the absolute path to the entrypoint executable
+    uint32_t path_length = 0;
+    if (_NSGetExecutablePath(NULL, &path_length) != -1)
+    {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    char path_buf[path_length];
+    if (_NSGetExecutablePath(path_buf, &path_length) != 0)
+    {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    return realpath(path_buf, NULL);
```

Comment:
Just curious about the NULL second argument here.  I'm assuming it works, as otherwise CI would be falling over, but in the OSX man page it doesn't call out NULL specially and instead says:
https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/realpath.3.html
> The resolved_name argument must refer to a buffer capable of storing at least PATH_MAX characters.

whereas the Linux man page says:

https://www.man7.org/linux/man-pages/man3/realpath.3.html
> If resolved_path is specified as NULL, then realpath() uses malloc(3) to allocate a buffer of up to PATH_MAX bytes to hold the resolved pathname, and returns a pointer to this buffer.  The caller should deallocate this buffer using free(3).

Reply:
The [old POSIX](https://pubs.opengroup.org/onlinepubs/009695399/functions/realpath.html) described the behavior with NULL as implementation defined.

The [latest POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html) describes the behavior that we depend on here.

We are unlikely to ever target old system where this does not behave as expected. And if we do, there are multiple places in the repo that would need to be updated.

Reply:
+1, and illumos [manpage](https://illumos.org/man/3C/realpath) mentions a specific error:

```
       EINVAL
                       Either the file_name or resolved_name argument is a null
                       pointer.
```

in case getexecname() returns a non-absolute path, we get a segmentation fault (tested with `const char* getexecname_dummy() { return "foo/bar"; }`). However, if the `getexecname()` returned an absolute path (normal case), then realpath just short-circuits (ignoring the `resolved_name == NULL` case).

Reply:
We can prepend cwd for __sun like done in other places or `if (*path_buf != '/') { return null; }`.

Reply:
We have at least 3 other places in the repo that call realpath with NULL. I do not think it makes sense to do one-off fix here, without fixing the other places.

Just curious: How many libraries tests fail on Illumos? Does anybody use .NET runtime for anything real on Illumos?

Reply:
Current implementation here does not seem to have a huge issue as [`getexecname`](https://illumos.org/man/3c/getexecname)'s manpage mentions "Normally this is an absolute pathname" and `realpath(path,NULL)` just returns `path` for absolute path. It was just that in other places, we have __sun case handled like this: https://github.com/dotnet/runtime/blob/cd4cc97e4c099f637061afe2b6c546483ffd3073/src/installer/corehost/cli/hostmisc/pal.unix.cpp#L868

here, since we are handling __sun case, using `if (*path_buf != '/') { /* return null or prepend cwd */ }` before `return realpath` will avoid sigsegv, if we ever hit the non-normal case.